### PR TITLE
Add doc for Java client shadowing

### PR DESCRIPTION
**Pull request by yux0**
Add doc for Java client shadowing

### CODE REVIEW

**Review by yycptt — COMMENTED** [review]
I think you'll also need to update the config file: https://github.com/uber/cadence-docs/blob/master/src/.vuepress/config.js#L146

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+}
+```
+
+### Shadowing Worker
+
+NOTE: 
+- **All shadow workflows are running in one Cadence system domain, and right now, every user domain can only have one shadow workflow at a time.**
+- **The Cadence server used for scanning and getting workflow history will also be the Cadence server for running your shadow workflow.** Currently, there's no way to specify different Cadence servers for hosting the shadowing workflow and scanning/fetching workflow.
+
+Your worker can also be configured to run in shadow mode to run shadow tests as a workflow. This is useful if there's a number of workflows need to be replayed. Using a workflow can make sure the shadowing won't accidentally fail in the middle and the replay load can be distributed by deploying more shadow mode workers. It can also be incorporated into your deployment process to make sure there's no failed replay checks before deploying your change to production workers.
```
Comment (vytautas-karpavicius) [inline src/docs/04-java-client/18-workflow-replay-shadowing.md:123]: This is useful if there's a number of workflows `that` need to be replayed

**Review by vytautas-karpavicius — COMMENTED** [review]
I've noticed a few syntax errors while reading. Maybe it would be worth passing it through syntax checker?

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  public void testReplay() throws Exception {
+    WorkflowReplayer.replayWorkflowExecutionFromResource(
+        "HelloActivity.json", HelloActivity.GreetingWorkflowImpl.class);
+  }
+}
+```
+
+## Workflow Shadower
+
+Workflow Replayer works well when verifying the compatibility against a small number of workflow histories. If there are lots of workflows in production need to be verified, dumping all histories manually clearly won't work. Directly fetching histories from cadence server might be a solution, but the time to replay all workflow histories might be too long for a test.
```
[the remark this answers is not in the document]
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md:62]: Done

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+Workflow Shadower is built on top of Workflow Replayer to address this problem. The basic idea of shadowing is: scan workflows based on the filters you defined, fetch history for each of workflow in the scan result from Cadence server and run the replay test. It can be run either as a test to serve local development purpose or as a workflow in your worker to continuously replay production workflows.
```
[the remark this answers is not in the document]
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md:64]: done

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+- WorkflowQuery: If you are familiar with our advanced visibility query syntax, you can specify a query directly. If specified, all other scan filters must be left empty.
+- WorkflowTypes: A list of workflow Type names.
+- WorkflowStatuses: A list of workflow status. 
+- WorkflowStartTimeFilter: Min and max timestamp for workflow start time.
+- WorkflowSamplingRate: Sampling workflows from the scan result before executing the replay test.
+
+#### Shadow Exit Condition
+
+- ExpirationInterval: Shadowing will exit when the specified interval has passed.
+- ShadowCount: Shadowing will exit after this number of workflow has been replayed. Note: replay maybe skipped due to errors like can't fetch history, history too short, etc. Skipped workflows won't be taken account into ShadowCount.
```
[the remark this answers is not in the document]
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md:81]: done

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+#### Shadow Mode
+
+- Normal: Shadowing will complete after all workflows matches WorkflowQuery (after sampling) have been replayed or when exit condition is met.
+- Continuous: A new round of shadowing will be started after all workflows matches WorkflowQuery have been replayed. There will be a 5 min wait period between each round, and currently this wait period is not configurable. Shadowing will complete only when ExitCondition is met. ExitCondition must be specified when using this mode.
+
+#### Shadow Concurrency
+
+- Concurrency: workflow replay concurrency. If not specified, will be default to 1. For local shadowing, an error will be returned if a value higher than 1 is specified.
```
[the remark this answers is not in the document]
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md:90]: done

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+### Local Shadowing Test
+
+Local shadowing test is similar to the replay test. First create a workflow shadower with optional shadow and replay options, then register the workflow that need to be shadowed. Finally, call the `Run` method to start the shadowing. The method will return if shadowing has finished or any non-deterministic error is found.
```
[the remark this answers is not in the document]
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md:94]: done

File: src/docs/04-java-client/18-workflow-replay-shadowing.md (hunk 1 above)
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md:123]: done

File: src/docs/04-java-client/18-workflow-replay-shadowing.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md:81]: done

**Comment by yux0** [comment]
> I think you'll also need to update the config file: https://github.com/uber/cadence-docs/blob/master/src/.vuepress/config.js#L146

Good catch. I updated it

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+WorkflowReplayer.replayWorkflowExecution(history, MyWorkflowImpl.class);
+
+// if workflow history is stored in a json file
+WorkflowReplayer.replayWorkflowExecutionFromResource("workflowHistory.json", MyWorkflowImpl.class);
+
+// if workflow history is read from a File
+WorkflowReplayer.replayWorkflowExecution(historyFileObject, MyWorkflowImpl.class);
+```
+
+#### Step 5: Catch returned exception
```
Comment (yycptt) [inline src/docs/04-java-client/18-workflow-replay-shadowing.md:42]: step 3?

File: src/docs/04-java-client/18-workflow-replay-shadowing.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+When running in shadow mode, the normal decision worker will be disabled so that it won't update any production workflows. A special shadow activity worker will be started to execute activities for scanning and replaying workflows. The actual shadow workflow logic is controlled by Cadence server and your worker is only responsible for scanning and replaying workflows. 
+
+[Replay succeed, skipped and failed metrics](https://github.com/uber/cadence-java-client/blob/master/src/main/java/com/uber/cadence/internal/metrics/MetricsType.java#L169-L172) will be emitted by your worker when executing the shadow workflow and you can monitor those metrics to see if there's any incompatible changes. 
+
+To enable the shadow mode, the only change needed is setting the `EnableShadowWorker` field in `worker.Options` to `true`, and then specify the `ShadowOptions`.
```
Comment (yycptt) [inline src/docs/04-java-client/18-workflow-replay-shadowing.md:129]: Is it true for java client? For go client, only config needs to be changed. For java I thought there's a separate API for creating shadow worker?

Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md]: Yes. There is a separate class for it. I updated the doc here.

File: src/docs/04-java-client/18-workflow-replay-shadowing.md (hunk 7 above)
Comment (yux0) [reply src/docs/04-java-client/18-workflow-replay-shadowing.md]: Updated

**Review by yycptt — APPROVED** [review]

**Pull request merged**